File: LeetCode/1871. Jump Game VII/solution.py
class Solution:
    def canReach(self, s: str, minJump: int, maxJump: int) -> bool:
        n = len(s)
        if s[n-1] == '1':
            return False
        q = deque()
        q.append(0)
        reach = 0
        while q:
            cur = q.popleft()
            if cur == n-1:
                return True
            mi = max(reach+1,cur+minJump)
            mx = min(n,cur+maxJump+1)
            for nxt in range(mi, mx):
                if s[nxt] == '0':
                    q.append(nxt)
                reach = max(reach, nxt)
        return False

File: LeetCode/1871. Jump Game VII/solution.cpp
class Solution {
public:
    bool canReach(string s, int minJump, int maxJump) {
        int n = s.length();
        if (s[n-1] == '1')
            return false;
        s[0] = '#';
        int reach = 0;
        for (int i = 0; i < n; i++){
            if (s[i] == '#'){
                reach = max(reach, i+minJump);
                while (reach < n && reach <= i+maxJump){
                    if (s[reach] == '0')
                        s[reach] = '#';
                    reach++;
                }
            }
        }
        return s[n-1] == '#';
    }
};